### ISSUE TITLE
[JIT]: JIT not working properly. like flex classes (next js)

### ISSUE THREAD

**Issue by MuttakinHasib** [tailwindlabs/tailwindcss#4167]
### What version of Tailwind CSS are you using?

v2.1.1

### What build tool (or framework if it abstracts the build tool) are you using?

next.js 10.1.3

### What version of Node.js are you using?

v14.15.1

### What browser are you using?

Firefox, Chrome

### What operating system are you using?

macOS

### Reproduction repository

https://github.com/MuttakinHasib/tealeaf

### Describe your issue

JIT not working properly... when i use classs like w-20 or flex classes its not working properly
@adamwathan  can you check my repo ? is it bug or i missed something ?

**Comment by joshxyzhimself** [comment]
Hi, could this possibly be addressed in release v2.1.2?

https://github.com/tailwindlabs/tailwindcss/releases/tag/v2.1.2

**Comment by adamwathan** [comment]
Hey! Cloned and I can't reproduce any issues here, flex classes work as does w-20:

<img width="1242" alt="image" src="https://user-images.githubusercontent.com/4323180/115991148-356d5580-a595-11eb-8668-a684315bfbeb.png">

<img width="1242" alt="image" src="https://user-images.githubusercontent.com/4323180/115991149-37cfaf80-a595-11eb-9e1a-a81f58210855.png">

Going to close but if this is still an issue please create a new issue with very specific steps to follow to see the issues you are seeing.

**Comment by SleeplessByte** [comment]
With `"tailwindcss": "^2.1.2"` and the `tailwind.config.js` that looks like this:

```javascript
module.exports = {
  mode: 'jit',
  future: {
    removeDeprecatedGapUtilities: true,
    purgeLayersByDefault: true,
  },
  purge: [
    './public/**/*.html',
    './components/**/*.{js,ts,jsx,tsx}',
    './pages/**/*.{js,ts,jsx,tsx}',
  ],
  variants: {},
}
```

...and running `"dev": "next dev -p 8081"`...

I experience the exact same issue as @MuttakinHasib.

```jsx
<section className="px-12 mt-10 flex flex-col" style={{ minHeight: 410 }}>
  <ul className="mt-5 flex flex-1">
    <li className="w-24">
    </li>
  </ul>
</section>
```

- [x] `px-12` works 
- [x] `mt-10` works
- [ ] `px-10` doesn't work (not shown here, but observed when changing classname)
- [ ] `my-10` doesn't either (not shown here, but observed when changing classname)
- [x] `flex` works
- [ ] `flex-col` doesn't work
- [x] `mt-5` works
- [x] `flex-1` works
- [ ] `w-24` doesn't work

It seems sorta consistent. Restarting the server doesn't resolve the issues. 

Workaround is to run `rm .next/cache -rf`, and then restart the server. Then it works. This makes me assume that the cause isn't necessarily in JIT, but in how JIT output is picked up by next (or not picked up) or how the cache is breaking it.

**Comment by lednhatkhanh** [comment]
I'm also facing the exact same issue

**Comment by aludvigsen** [comment]
Im consistently experiencing the same issue, _unconsistently_. A restart after clearing the `/.next` cache always resolves it. I dont know why, maybe it is an Nextjs issue,

**Comment by MuttakinHasib** [comment]
@aludvigsen  yeah... same here 
@adamwathan

**Comment by adamwathan** [comment]
Please open a new issue with a reproduction ❤️

**Comment by SleeplessByte** [comment]
I can't give you a consistent reproduction, so I won't open a new issue for _the same problem_ that @MuttakinHasib reported and was closed because 'it couldn't be reproduced'. You'll just close that issue again with a "can't reproduce", and that would make sense :).

My mentioned workaround works **consistently**, but that should further indicate that this is a valid issue.

If someone does find a way to consistently reproduce this, that would be very helpful for everyone.

**Comment by adamwathan** [comment]
Can you folks try setting `TAILWIND_DISABLE_TOUCH=true` on the command line before starting your dev server? It'll use a different dependency tracking system that should be more resilient to how the persistent file cache in Webpack 5 works.

It would still be better if someone could create a minimal reproduction just so I know how it's all set up and can think through the possible reasons this problem could occur. It doesn't need to demonstrate the issue immediately, it's fine if the reproduction steps are "edit templates in the project for an hour and notice that eventually certain changes stop being picked up but it's different every time". It would just be helpful to see what version of webpack, what version of Next, how the purge paths are configured, what flags are being passed to Tailwind through environment variables, etc.

**Comment by sandipndev** [comment]
I also noticed the same issue when working with Next.js -e `with-tailwindcss` which set tailwind's mode as "jit".
After deleting the next cache, it works.

**Comment by Zertz** [comment]
I know this is hardly minimal but here goes.

I'm running into this issue in a yarn workspace with Next.js in a subdirectory. I'm not sure you'll be able to reproduce it right after a clone but it's happening for me locally. I noticed when it wouldn't generate `text-3xl` and `-mb-4` but it looks like a caching thing so I don't think the missing classes actually matter.

https://github.com/Zertz/react-headless-tabs/tree/v2-rc

1. Run `yarn` in the root
2. `cd docs`
3. `yarn dev`
4. 🤞

**Comment by Elia-Darwish** [comment]
I’m also experiencing the same regularly on both dev and production environments. I actually am always removing the .next directory before rebuilding the project to ensure the style changes are picked up.

What I wanted to add up is that I noticed this happening mostly when changing the value of the same utility repeatedly and sometimes when using arbitrary values as well.

This is of course all theoretical, but hopefully will help a little bit 😅

cheers

**Comment by mohux** [comment]
Same thing happens, sadly am facing it that I have to clear the cache everytime i update a value, never happened before 

with tailwind 2.1.2

**Comment by unigazer** [comment]
I've might found a potential fix, but I haven't fully tested it yet.

- Put components, styles, etc., and pages in the "src" folder, at root.
- Add `mode: 'jit'` in `tailwind.config.js`

I have followed this article: https://blog.devsharma.live/how-to-setup-tailwind-css-jit-with-nextjs

My opinion that it's not a problem in Tailwind, but rather Next.js itself, but I'm not certain of it.

**Comment by octokatherine** [comment]
I am running into the same issue, but even clearing the cache isn't fixing it

**Comment by adamwathan** [comment]
Hey @octokatherine, can you open a new issue and provide a reproduction? If we can see it not working on our own machines we can figure it out way, way faster.

Guesses:

- Make sure on absolute latest version of everything (Next 11+, Tailwind 2.2+)
- Make sure no mistakes in your `purge` config
- Make sure not using string interpolation in your class names

**Comment by unigazer** [comment]
If your `jit` mode isn't working at all, try to structure your project like the one [from this article](https://blog.devsharma.live/how-to-setup-tailwind-css-jit-with-nextjs). My issue was that `jit` wasn't working at all, but after reorganizing the project structure by following the linked article, now it works all the time.

**Comment by hussamkhatib** [comment]
> Im consistently experiencing the same issue, _unconsistently_. A restart after clearing the `/.next` cache always resolves it. I dont know why, maybe it is an Nextjs issue,

Do you know how we can clear cache in windows

**Comment by SleeplessByte** [comment]
Stop the server. Yes, delete `.next/cache` in your project folder.

**Comment by sabit990928** [comment]
> Hey @octokatherine, can you open a new issue and provide a reproduction? If we can see it not working on our own machines we can figure it out way, way faster.
> 
> Guesses:
> 
>     * Make sure on absolute latest version of everything (Next 11+, Tailwind 2.2+)
> 
>     * Make sure no mistakes in your `purge` config
> 
>     * Make sure not using string interpolation in your class names

I already used latest versions of Tailwind css, but upgrading the Next.js version from 10.1.3 to 11.0.1 solved the issue for me. Thanks @adamwathan 

JIT is amazing)

**Comment by jellysquider** [comment]
Hello! I've encountered this issue today as well. I noticed that when I put tailwind directly into `src/pages` it suddenly works but then just as suddenly stops working on the next server start. I'm not sure what's going on 

Here's a super basic example with 2 navbars (one shown on mobile, one on desktop). None of the properties work for me for some reason

**Comment by Manas-E** [comment]
Hello @adamwathan 

I've also faced the same issue and tried everything but nothing worked, interestingly by doing some changes in tailwind.config file it was working perfectly but only for once.

So I tried **updating react and next.js** to the latest version and Voila !! it worked, thanks for the suggestion.

Anybody facing the same issue just try updating your react and next to the latest and clean the cache by doing this " rm .next/cache -rf"

**Comment by tengkuzulfadli** [comment]
Hello @jellysquider ,

I did follow your steps but it's still not working.

The issue is, when I start using/declaring for xs / sm screen, it will change all styles for other screen sizes.

For example:

<h3 className="text-center xs:text-left">Explore our Range</h3>

It will change the text to the left for all screen! I tried to override using plain css, it didn't work!

The purge inside tailwind.config.js is as below:

purge: ['./src/pages/**/*.{js,ts,jsx,tsx}', './src/components/**/*.{js,ts,jsx,tsx}'],

My folders structure is similar to yours. I also cleared all next/cache

Do you have any idea?

**Comment by odafeaggreh** [comment]
> rm .next/cache -rf



> 

This worked for me thank you, though I only had to upgrade **Tailwind** by running: `npm install -D tailwindcss@latest postcss@latest autoprefixer@latest`  and not nextjs after which I cleared the cache by running: `rm -r .next/cache`

I hope this helps someone 🙂

**Comment by SleeplessByte** [comment]
On the majority of systems where either the build tool doesn't support this yet, using ` TAILWIND_MODE=watch some-run-command`  solves the issue.
